akka-typed/src/main/scala/akka/typed/cluster/ClusterSharding.scala

```diff
@@ -0,0 +1,196 @@
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class DefaultNoEnvelopeMessageExtractor[A](maxNumberOfShards: Int) extends TypedMessageExtractor[A, A] {
+  def entityMessage(message: A) = message
+  def shardId(message: A) = {
+    val id = entityId(message)
+    if (id != null) (math.abs(id.hashCode) % maxNumberOfShards).toString
+    else null
+  }
+}
+
+/**
+ * A reference to an entityId and the local access to sharding, allows for actor-like interaction
+ *
+ * The entity ref must be resolved locally and cannot be sent to another node.
+ *
+ * TODO what about ask, should it actually implement ActorRef to be exactly like ActorRef and callers does not have
+ *      to know at all about it or is it good with a distinction but lookalike API?
+ */
+trait EntityRef[A] {
+  /**
+   * Send a message to the entity referenced by this EntityRef using *at-most-once*
+   * messaging semantics.
+   */
+  def tell(msg: A): Unit
+}
+
+object EntityRef {
+  implicit final class EntityRefOps[T](val ref: EntityRef[T]) extends AnyVal {
+    /**
+     * Send a message to the Actor referenced by this ActorRef using *at-most-once*
+     * messaging semantics.
+     */
+    def !(msg: T): Unit = ref.tell(msg)
+  }
+}
+
+object ClusterSharding extends ExtensionId[ClusterSharding] {
+  def createExtension(system: ActorSystem[_]): ClusterSharding = ???
+
+  /**
+   * Create an ActorRef-like reference to a specific sharded entity. Messages sent to it will be wrapped
+   * in a [[ShardingEnvelope]] and passed to the local shard region or proxy.
+   */
+  def entityRefFor[A](entityId: String, actorRef: ActorRef[ShardingEnvelope[A]]): EntityRef[A] =
+    new EntityRef[A] {
+      def tell(msg: A): Unit = actorRef ! ShardingEnvelope(entityId, msg)
+    }
+
+}
+
+trait ClusterSharding extends Extension {
+
+  /**
+   * Spawn a shard region or a proxy depending on if the settings require role and if this node has such a role.
+   *
+   * Messages are sent to the entities by wrapping the messages in a [[ShardingEnvelope]] with the entityId of the
+   * recipient actor.
+   * A [[DefaultMessageExtractor]] will be used for extracting entityId and shardId
+   * [[akka.cluster.sharding.ShardCoordinator.LeastShardAllocationStrategy]] will be used for shard allocation strategy.
+   *
+   * @param behavior The behavior for entities
+   * @param typeName A name that uniquely identifies the type of entity in this cluster
+   * @param handOffStopMessage Message sent to an entity to tell it to stop
+   * @tparam A The type of command the entity accepts
+   */
+  def spawn[A](
+    behavior:           Behavior[A],
+    typeName:           String,
+    settings:           ClusterShardingSettings,
+    maxNumberOfShards:  Int,
+    handOffStopMessage: A): ActorRef[ShardingEnvelope[A]]
```

Comment:
allocationStrategy, we might have some overloads with default values also

Reply:
role and dataCenter are also missing

Reply:
I went for one simplest possible and one with all the bells and whistles, guess I missed a few whistles though. 

Maybe fold more things into settings, or provide a builder-type and a single API entry point instead of many different overloads?